From 2c173061b17b861b51dfbbe60e361da5ec7de057 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Tue, 2 Jul 2024 10:43:20 +0200
Subject: [PATCH 1091/1695] drivers: clk: fix error cases in STM32MP21 clocks

Fix missing test on some function return code in stm32mp21 clock
driver. Most of these test are required, few are not strictly needed
since the called function always return with success but even so it
is preferred to test the return value for implementation consistency
and maintenance.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 8e60ffdca5084003d56b05aff551a2308c7f4e32 drivers: clk: fix error cases in STM32MP21 clocks]
Change-Id: Ib0eb238348ba571797078bb79b82dd6362987246
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/390721
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Gabriel FERNANDEZ <gabriel.fernandez@foss.st.com>
---
 core/drivers/clk/clk-stm32mp21.c | 41 ++++++++++++++++++++------------
 1 file changed, 26 insertions(+), 15 deletions(-)

diff --git a/core/drivers/clk/clk-stm32mp21.c b/core/drivers/clk/clk-stm32mp21.c
index bf2c85da3..d40e70ed4 100644
--- a/core/drivers/clk/clk-stm32mp21.c
+++ b/core/drivers/clk/clk-stm32mp21.c
@@ -1787,7 +1787,8 @@ static int clk_stm32_pll_init(struct clk_stm32_priv *priv, int pll_idx,
 	 * a configuration on the fly.
 	 */
 
-	stm32_gate_rdy_disable(pll->gate_id);
+	if (stm32_gate_rdy_disable(pll->gate_id))
+		panic();
 
 	ret = clk_stm32_pll_set_mux(priv, pll_conf->src);
 	if (ret != 0)
@@ -1805,7 +1806,8 @@ static int clk_stm32_pll_init(struct clk_stm32_priv *priv, int pll_idx,
 		spread_spectrum = true;
 	}
 
-	stm32_gate_rdy_enable(pll->gate_id);
+	if (stm32_gate_rdy_enable(pll->gate_id))
+		panic();
 
 	if (spread_spectrum)
 		io_clrbits32(pllxcfgr1, RCC_PLLxCFGR1_SSMODRST);
@@ -2238,8 +2240,8 @@ static void __maybe_unused clk_stm32_osc_pm_restore(struct clk *clk)
 	if (!stm32_rcc_has_access_by_id(RCC_RIF_OSCILLATORS))
 		return;
 
-	if (clk_is_enabled(clk))
-		clk_stm32_osc_enable(clk);
+	if (clk_is_enabled(clk) && clk_stm32_osc_enable(clk))
+		panic();
 }
 
 static const struct clk_ops clk_stm32_osc_ops = {
@@ -2266,8 +2268,8 @@ static void __maybe_unused clk_stm32_osc_msi_pm_restore(struct clk *clk)
 	if (!stm32_rcc_has_access_by_id(RCC_RIF_OSCILLATORS))
 		return;
 
-	if (clk_is_enabled(clk))
-		clk_stm32_osc_enable(clk);
+	if (clk_is_enabled(clk) && clk_stm32_osc_enable(clk))
+		panic();
 }
 
 static const struct clk_ops clk_stm32_oscillator_msi_ops = {
@@ -2292,7 +2294,8 @@ static void __maybe_unused clk_stm32_hse_div_pm_restore(struct clk *clk)
 	if (!stm32_rcc_has_access_by_id(RCC_RIF_OSCILLATORS))
 		return;
 
-	clk_stm32_hse_div_set_rate(clk, clk->rate, clk->parent->rate);
+	if (clk_stm32_hse_div_set_rate(clk, clk->rate, clk->parent->rate))
+		panic();
 }
 
 static const struct clk_ops clk_stm32_hse_div_ops = {
@@ -2305,7 +2308,7 @@ static const struct clk_ops clk_stm32_hse_div_ops = {
 static TEE_Result clk_stm32_hsediv2_enable(struct clk *clk)
 {
 	if (stm32_rcc_has_access_by_id(RCC_RIF_OSCILLATORS))
-		clk_stm32_gate_enable(clk);
+		return clk_stm32_gate_enable(clk);
 
 	return TEE_SUCCESS;
 }
@@ -2515,10 +2518,16 @@ static TEE_Result clk_stm32_pll_enable(struct clk *clk)
 {
 	struct clk_stm32_pll_cfg *cfg = clk->priv;
 
-	if (stm32_rcc_has_access_by_id(RCC_RIF_PLL4_TO_8))
-		if (stm32_gate_rdy_enable(cfg->gate_id) != 0U)
+	if (stm32_rcc_has_access_by_id(RCC_RIF_PLL4_TO_8)) {
+		TEE_Result res = TEE_ERROR_GENERIC;
+
+		res = stm32_gate_rdy_enable(cfg->gate_id);
+		if (res)
 			EMSG("%s timeout", clk_get_name(clk));
 
+		return res;
+	}
+
 	return TEE_SUCCESS;
 }
 
@@ -2543,8 +2552,8 @@ static void __maybe_unused clk_stm32_pll_pm_restore(struct clk *clk)
 	if (!stm32_rcc_has_access_by_id(RCC_RIF_PLL4_TO_8))
 		return;
 
-	if (clk_is_enabled(clk))
-		clk_stm32_pll_enable(clk);
+	if (clk_is_enabled(clk) && clk_stm32_pll_enable(clk))
+		panic();
 }
 
 static const struct clk_ops clk_stm32_pll_ops = {
@@ -3010,11 +3019,13 @@ static void __maybe_unused clk_stm32_rif_composite_pm_restore(struct clk *clk)
 		if (clk_get_parent_idx(clk, clk->parent, &pidx))
 			panic();
 
-		stm32_mux_set_parent(cfg->mux_id, pidx);
+		if (stm32_mux_set_parent(cfg->mux_id, pidx))
+			panic();
 	}
 
-	if (cfg->div_id != NO_DIV)
-		stm32_div_set_rate(cfg->div_id, clk->rate, clk->parent->rate);
+	if (cfg->div_id != NO_DIV &&
+	    stm32_div_set_rate(cfg->div_id, clk->rate, clk->parent->rate))
+		panic();
 
 	if (cfg->gate_id != NO_GATE)
 		if (clk_is_enabled(clk))
-- 
2.47.2

